lightningd/plugin_control.c

```diff
@@ -286,28 +311,23 @@ static struct command_result *json_plugin_control(struct command *cmd,
 			   NULL))
 			return command_param_failed();
 
-		plugins_add_default_dir(cmd->ld->plugins,
-				path_join(tmpctx, cmd->ld->config_dir, "plugins"));
+		/* Delay to avoid a race condition with command_still_pending */
+		new_reltimer(cmd->ld->timers, cmd,
+		             time_from_sec(1),
+		             plugin_dynamic_rescan_plugins, cmd);
```

Comment:
Why ? I thought that was because `command_still_pending` was executed after the `command_success`.

Reply:
I think @darosior is correct.  I had to do this a number of times long ago in my `payalgo.c` predecessor to the `pay` plugin.  Caller of some utility function (which expects a callback rather than returns directly itself) calls `command_still_pending`, then we need to trigger `command_success` inside the callback, and the utility function is able to determine this condition (need to call the callback) immediately without deferring the callback.  Only way to get around that was to pointlessly delay.

Though it should be fine to use `time_from_sec(0)` in that case instead of pointlessly delaying by 1 second.

Reply:
I experienced this too, but it usually happens because you haven't plumbed through the command_result.

This seems to work:

```
diff --git a/lightningd/plugin_control.c b/lightningd/plugin_control.c
index 44a56782c..857e0b891 100644
--- a/lightningd/plugin_control.c
+++ b/lightningd/plugin_control.c
@@ -10,7 +10,7 @@ struct dynamic_plugin {
 /**
  * Returned by all subcommands on success.
  */
-static void plugin_dynamic_list_plugins(struct command *cmd)
+static struct command_result *plugin_dynamic_list_plugins(struct command *cmd)
 {
 	struct json_stream *response;
 	struct plugin *p;
@@ -25,22 +25,23 @@ static void plugin_dynamic_list_plugins(struct command *cmd)
 		json_object_end(response);
 	}
 	json_array_end(response);
-	was_pending(command_success(cmd, response));
+	return command_success(cmd, response);
 }
 
 /**
  * Returned by all subcommands on error.
  */
-static void plugin_dynamic_error(struct dynamic_plugin *dp, const char *error)
+static struct command_result *plugin_dynamic_error(struct dynamic_plugin *dp,
+						   const char *error)
 {
 	plugin_kill(dp->plugin, "%s: %s", dp->plugin->cmd, error);
-	was_pending(command_fail(dp->cmd, JSONRPC2_INVALID_PARAMS,
-	                         "%s: %s", dp->plugin->cmd, error));
+	return command_fail(dp->cmd, JSONRPC2_INVALID_PARAMS,
+			    "%s: %s", dp->plugin->cmd, error);
 }
 
 static void plugin_dynamic_timeout(struct dynamic_plugin *dp)
 {
-	plugin_dynamic_error(dp, "Timed out while waiting for plugin response");
+	was_pending(plugin_dynamic_error(dp, "Timed out while waiting for plugin response"));
 }
 
 static void plugin_dynamic_config_callback(const char *buffer,
@@ -61,7 +62,7 @@ static void plugin_dynamic_config_callback(const char *buffer,
 	}
 
 	/* No plugin unconfigured left, return the plugin list */
-	plugin_dynamic_list_plugins(dp->cmd);
+	was_pending(plugin_dynamic_list_plugins(dp->cmd));
 }
 
 /**
@@ -85,11 +86,15 @@ static void plugin_dynamic_manifest_callback(const char *buffer,
                                              const jsmntok_t *idtok,
                                              struct dynamic_plugin *dp)
 {
-	if (!plugin_parse_getmanifest_response(buffer, toks, idtok, dp->plugin))
-		return plugin_dynamic_error(dp, "Gave a bad response to getmanifest");
+	if (!plugin_parse_getmanifest_response(buffer, toks, idtok, dp->plugin)) {
+		was_pending(plugin_dynamic_error(dp, "Gave a bad response to getmanifest"));
+		return;
+	}
 
-	if (!dp->plugin->dynamic)
-		return plugin_dynamic_error(dp, "Not a dynamic plugin");
+	if (!dp->plugin->dynamic) {
+		was_pending(plugin_dynamic_error(dp, "Not a dynamic plugin"));
+		return;
+	}
 
 	/* We got the manifest, now send the init message */
 	plugin_dynamic_config(dp);
@@ -99,7 +104,7 @@ static void plugin_dynamic_manifest_callback(const char *buffer,
  * This starts a plugin : spawns the process, connect its stdout and stdin,
  * then sends it a getmanifest request.
  */
-static void plugin_start(struct dynamic_plugin *dp)
+static struct command_result *plugin_start(struct dynamic_plugin *dp)
 {
 	int stdin, stdout;
 	char **p_cmd;
@@ -131,32 +136,33 @@ static void plugin_start(struct dynamic_plugin *dp)
 	                            plugin_dynamic_manifest_callback, dp);
 	jsonrpc_request_end(req);
 	plugin_request_send(p, req);
+	return command_still_pending(dp->cmd);
 }
 
 /**
  * Called when trying to start a plugin through RPC, it starts the plugin and
  * will give a result 20 seconds later at the most.
  */
-static void plugin_dynamic_start(struct command *cmd, const char *plugin_path)
+static struct command_result *plugin_dynamic_start(struct command *cmd,
+						   const char *plugin_path)
 {
 	struct dynamic_plugin *dp;
 
 	dp = tal(cmd, struct dynamic_plugin);
 	dp->cmd = cmd;
 	dp->plugin = plugin_register(cmd->ld->plugins, plugin_path);
-	if (!dp->plugin) {
-		plugin_dynamic_error(dp, "Is already registered");
-		return;
-	}
+	if (!dp->plugin)
+		return plugin_dynamic_error(dp, "Is already registered");
 
-	plugin_start(dp);
+	return plugin_start(dp);
 }
 
 /**
  * Called when trying to start a plugin directory through RPC, it registers
  * all contained plugins recursively and then starts them.
  */
-static void plugin_dynamic_startdir(struct command *cmd, const char *dir_path)
+static struct command_result *plugin_dynamic_startdir(struct command *cmd,
+						      const char *dir_path)
 {
 	const char *err;
 	struct plugin *p;
@@ -165,7 +171,7 @@ static void plugin_dynamic_startdir(struct command *cmd, const char *dir_path)
 
 	err = add_plugin_dir(cmd->ld->plugins, dir_path, false);
 	if (err)
-		return was_pending(command_fail(cmd, JSONRPC2_INVALID_PARAMS, "%s", err));
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS, "%s", err);
 
 	found = false;
 	list_for_each(&cmd->ld->plugins->plugins, p, list) {
@@ -178,7 +184,9 @@ static void plugin_dynamic_startdir(struct command *cmd, const char *dir_path)
 		}
 	}
 	if (!found)
-		plugin_dynamic_list_plugins(cmd);
+		return plugin_dynamic_list_plugins(cmd);
+	else
+		return command_still_pending(cmd);
 }
 
 static struct command_result *
@@ -211,7 +219,7 @@ plugin_dynamic_stop(struct command *cmd, const char *plugin_name)
 /**
  * Look for additions in the default plugin directory.
  */
-static void plugin_dynamic_rescan_plugins(struct command *cmd)
+static struct command_result *plugin_dynamic_rescan_plugins(struct command *cmd)
 {
 	bool found;
 	struct plugin *p;
@@ -231,7 +239,9 @@ static void plugin_dynamic_rescan_plugins(struct command *cmd)
 	}
 
 	if (!found)
-		plugin_dynamic_list_plugins(cmd);
+		return plugin_dynamic_list_plugins(cmd);
+
+	return command_still_pending(cmd);
 }
 
 /**
@@ -270,7 +280,7 @@ static struct command_result *json_plugin_control(struct command *cmd,
 			return command_param_failed();
 
 		if (access(plugin_path, X_OK) == 0)
-			plugin_dynamic_start(cmd, plugin_path);
+			return plugin_dynamic_start(cmd, plugin_path);
 		else
 			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
 						   "%s is not executable: %s",
@@ -285,7 +295,7 @@ static struct command_result *json_plugin_control(struct command *cmd,
 			return command_param_failed();
 
 		if (access(dir_path, F_OK) == 0)
-			plugin_dynamic_startdir(cmd, dir_path);
+			return plugin_dynamic_startdir(cmd, dir_path);
 		else
 			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
 						   "Could not open %s", dir_path);
@@ -295,23 +305,17 @@ static struct command_result *json_plugin_control(struct command *cmd,
 			   NULL))
 			return command_param_failed();
 
-		/* Delay to avoid a race condition with command_still_pending */
-		new_reltimer(cmd->ld->timers, cmd,
-		             time_from_sec(0),
-		             plugin_dynamic_rescan_plugins, cmd);
+		return plugin_dynamic_rescan_plugins(cmd);
 	} else if (streq(subcmd, "list")) {
 		if (!param(cmd, buffer, params,
 			   p_req("subcommand", param_ignore, cmd),
 			   NULL))
 			return command_param_failed();
 
-		/* Delay to avoid a race condition with command_still_pending */
-		new_reltimer(cmd->ld->timers, cmd,
-		             time_from_sec(0),
-		             plugin_dynamic_list_plugins, cmd);
+		return plugin_dynamic_rescan_plugins(cmd);
 	}
-
-	return command_still_pending(cmd);
+	/* subcmd must be one of the above: param_subcommand checked it! */
+	abort();
 }
 
 static const struct json_command plugin_control_command = {
```

Reply:
Ok thank you !
Having the error and success utilities returning `command_result` makes more sense.
I hesitated to put the `abort()` as I've seen you did it for the `dev` command but wasnt sure..